content/en/docs/architecture/timeouts.md

```diff
@@ -0,0 +1,69 @@
+---
+title: "Job Execution Timeouts"
+date: 2020-12-21T23:27:39-04:00
+draft: false
+---
+
+A typical OpenShift CI job runs [ci-operator](/docs/architecture/ci-operator/) in a pod which for each [step](/docs/architecture/step-registry/#step) of the job creates a pod, executing the commands defined by `tests.steps.test.commands` of a step in the job. A _test process_ is a process which runs such commands. There are timeouts in various phases of a job execution. In this document, we will explain those timeouts, show the logs if they occur in a job, and how the test process will be notified, and how to modify the values of those timeouts.
+
+## Job Timeout
+
+The job timeout can be considered as the overall time for `ci-operator` to run. When a job fails due to the job timeout, the job log contains:
+
+```
+Process did not finish before 4h0m0s timeout
+```
+
+The following stanza in [Prow's configuration](https://github.com/openshift/release/blob/9238ee8b89c861793f9b8e8f7f6509abe33bd0b8/core-services/prow/02_config/_config.yaml#L559) specifies the default value of job timeout and it can be overwritten at `org`, or `org/repo` level (not supported at a more specific level).
+
+```yaml
+plank: # a core component of Prow
+  default_decoration_configs:
+    '*':
+      grace_period: 30m0s
+      timeout: 4h0m0s
+    'org/repo': # overwrite the job timeout at repo level
+      grace_period: 45m0s
+      timeout: 6h0m0s
+```
+
+When the job timeout appears, the job pod will be deleted. The `ci-operator` process in the pod will receive [SIGTERM](https://pkg.go.dev/syscall#SIGTERM) and start to delete the pods it creates. If `ci-operator` is still running after `grace_period`, which is also specified in Prow's config, it will be forcibly halted and the job log contains:
+
+```
+Process did not exit before 30m0s grace period
+```
+
+Similarly, the test process receives [SIGTERM](https://pkg.go.dev/syscall#SIGTERM) when the pod running it gets deleted. If the pod is for a step of the job, the test process has to ensure that all artifacts of the step have to be ready before `termination_grace_period_seconds` (see [Step Timeout](/docs/architecture/timeouts/#step-timeout)) specified in the step definition is up.
```

Comment:
I think I'd phrase this sentence slightly differently, because the document is suppose to be about timeouts, and the mention about receiving SIGTERM comes a bit out of nowhere. Something like:

"Same grace period will also apply when the pod is terminated due to a different reason than a timeout, such as when  the `Pod` is deleted from the cluster" or something similar.

Reply:
Updated.

I was trying to address what happens to the test process (receive `SIGTERM`) while timeout takes place, and what it need to do (e.g., finishing on producing artifacts), and how much time it has before being killed (`30s` for a non-step job) and configurable for a step job).